### PR TITLE
[BEAM-2730] make BeamRecord an immutable type

### PR DESCRIPTION
**Pull request by XuMingmin**
@robertwb 

Ps, some lines are expected to delete with #3691 and #3689

### CODE REVIEW

**Comment by XuMingmin** [comment]
R: + @takidau

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamValuesRel.java [hunk 1]
```diff
@@ -65,11 +65,12 @@ public BeamValuesRel(
 
     BeamSqlRecordType beamSQLRowType = CalciteUtils.toBeamRowType(this.getRowType());
     for (ImmutableList<RexLiteral> tuple : tuples) {
-      BeamRecord row = new BeamRecord(beamSQLRowType);
+      List<Object> fieldsValue = new ArrayList<>();
```
Comment (takidau) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamValuesRel.java]: Same here, new ArrayList<>(tuple.size())?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/schema/BeamTableUtils.java [hunk 2]
```diff
@@ -39,7 +41,7 @@ public static BeamRecord csvLine2BeamSqlRow(
       CSVFormat csvFormat,
       String line,
       BeamSqlRecordType beamSqlRowType) {
-    BeamRecord row = new BeamRecord(beamSqlRowType);
+    List<Object> fieldsValue = new ArrayList<>();
```
Comment (takidau) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/schema/BeamTableUtils.java]: Same here using beamSqlRowType.size()?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/BeamRecordCoder.java [hunk 3]
```diff
@@ -73,14 +74,15 @@ public void encode(BeamRecord value, OutputStream outStream)
   public BeamRecord decode(InputStream inStream) throws CoderException, IOException {
     BitSet nullFields = nullListCoder.decode(inStream);
 
-    BeamRecord record = new BeamRecord(recordType);
+    List<Object> fieldValues = new ArrayList<>();
```
Comment (XuMingmin) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/BeamRecordCoder.java]: right, the size is known.

**Comment by takidau** [comment]
Merged, feel free to close. Thank you!

**Pull request closed without merging**